src/server.rs

```diff
@@ -237,8 +251,18 @@ impl Read for Encoder {
                     // it into the actual buffer
                     let mut chunk_buf = vec![0; buffer_remaining];
                     // Read bytes from body reader
-                    let chunk_length =
-                        ready!(Pin::new(&mut self.res).poll_read(cx, &mut chunk_buf))?;
+                    let inner_poll_result = Pin::new(&mut self.res).poll_read(cx, &mut chunk_buf);
+                    let chunk_length = match inner_poll_result {
+                        Poll::Ready(Ok(n)) => n,
+                        Poll::Pending => {
+                            if bytes_read == 0 {
+                                return Poll::Pending;
+                            } else {
+                                break;
+                            }
+                        }
+                        e => return e,
```

Comment:
Even though this is more typing, this should be easier to follow since all left-hand match arms then use fully qualified types, rather than a named catch-all argument as is the case now:

```suggestion
                        Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
```